common/raw_hashtable.h

```diff
@@ -1399,79 +1483,78 @@ BaseImpl<InputKeyT, InputValueT, InputKeyContextT>::GrowAndInsert(
+    if (&arg == this) {
+      return *this;
+    }
+    CARBON_DCHECK(arg.storage() != this->storage());
+    // The sizes match, so just clear out old entries.
+    if constexpr (!EntryT::IsTriviallyDestructible) {
+      this->view_impl_.ForEachEntry([](EntryT& entry) { entry.Destroy(); },
+                                    [](auto...) {});
     }
+  } else {
+    // The sizes don't match so destroy everything and re-setup the table
+    // storage.
+    this->Destroy();
+    this->alloc_size() = arg.alloc_size();
+    SetupStorage();
   }
+  this->growth_budget_ = arg.growth_budget_;
+  this->CopySlotsFrom(arg);
+  return *this;
 }
 
 // Puts the incoming table into a moved-from state that can be destroyed or
 // re-initialized but must not be used otherwise.
 template <typename InputBaseT, ssize_t SmallSize>
 TableImpl<InputBaseT, SmallSize>::TableImpl(TableImpl&& arg) noexcept
     : BaseT(arg.alloc_size(), arg.growth_budget_, SmallSize) {
+  CARBON_DCHECK(arg.alloc_size() > 0) << "Cannot move from a moved-from table!";
   CARBON_DCHECK(arg.small_alloc_size_ == SmallSize);
+  CARBON_DCHECK(!arg.is_small() || arg.alloc_size() == SmallSize);
+  CARBON_DCHECK(this->small_alloc_size_ == SmallSize);
+  this->MoveFrom(std::move(arg), small_storage());
+}
 
-  ssize_t local_size = arg.alloc_size();
-
-  if (SmallSize > 0 && arg.is_small()) {
-    CARBON_DCHECK(local_size == SmallSize);
-    this->storage() = small_storage();
+template <typename InputBaseT, ssize_t SmallSize>
+auto TableImpl<InputBaseT, SmallSize>::operator=(TableImpl&& arg) noexcept
+    -> TableImpl& {
+  CARBON_DCHECK(arg.alloc_size() > 0) << "Cannot move from a moved-from table!";
+  CARBON_DCHECK(arg.small_alloc_size_ == SmallSize);
+  CARBON_DCHECK(!arg.is_small() || arg.alloc_size() == SmallSize);
+  CARBON_DCHECK(this->small_alloc_size_ == SmallSize);
 
-    // For small tables, we have to move the entries as we can't move the tables
-    // themselves. We do this preserving their slots and even tombstones to
-    // avoid rehashing.
-    uint8_t* local_metadata = this->metadata();
-    EntryT* local_entries = this->entries();
-    uint8_t* local_arg_metadata = arg.metadata();
-    EntryT* local_arg_entries = arg.entries();
-    memcpy(local_metadata, local_arg_metadata, local_size);
-    if (EntryT::IsTriviallyRelocatable) {
-      memcpy(local_entries, local_arg_entries, SmallSize * sizeof(EntryT));
-    } else {
-      for (ssize_t group_index = 0; group_index < local_size;
-           group_index += GroupSize) {
-        auto g = MetadataGroup::Load(local_arg_metadata, group_index);
-        for (ssize_t byte_index : g.MatchPresent()) {
-          local_entries[group_index + byte_index].MoveFrom(
-              std::move(local_arg_entries[group_index + byte_index]));
-        }
-      }
-    }
-  } else {
-    // Just point to the allocated storage.
-    this->storage() = arg.storage();
-  }
+  // Destroy and deallocate our table.
+  this->Destroy();
```

Comment:
As far as I can tell, the latest LWG thinking is that self move assignment is valid but leaves the object in an unspecified state. I think this is enough to make that work:
```suggestion
  this->Destroy();
  this->alloc_size() = 0;
```

Reply:
Hmm, I think I was wrong here, and self move assignment is only supposed to work for a moved from object, for which I think this code already works.

Reply:
I had thought that self move mustn't be permitted at all because its already gone through `std::move`.... But now that you've mentioned it, that doesn't add up at all.

Even if the standard actually does't allow other forms of self move, I'm much more comfortable defending against it. Made equivalent change with a comment explaining things and a test.